stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/usecase/QueryMembers.kt

```diff
@@ -0,0 +1,76 @@
+package io.getstream.chat.android.livedata.usecase
+
+import io.getstream.chat.android.client.api.models.QuerySort
+import io.getstream.chat.android.client.call.Call
+import io.getstream.chat.android.client.call.CoroutineCall
+import io.getstream.chat.android.client.models.Member
+import io.getstream.chat.android.client.utils.FilterObject
+import io.getstream.chat.android.client.utils.Result
+import io.getstream.chat.android.livedata.ChatDomainImpl
+
+public class QueryMembers internal constructor(private val domainImpl: ChatDomainImpl) {
+
+    public operator fun invoke(
+        cid: String,
+        offset: Int = 0,
+        limit: Int = 0,
+        filter: FilterObject = FilterObject(),
+        sort: QuerySort<Member> = QuerySort.desc(Member::createdAt),
+        members: List<Member> = emptyList(),
+    ): Call<List<Member>> {
+        return CoroutineCall(domainImpl.scope) {
+            if (domainImpl.isOffline() && domainImpl.offlineEnabled) {
+                queryMembersOffline(cid, sort, offset, limit)
+            } else {
+                queryMembersOnline(cid, offset, limit, filter, sort, members)
+            }
+        }
+    }
+
+    private suspend fun queryMembersOnline(
+        cid: String,
+        offset: Int,
+        limit: Int,
+        filter: FilterObject,
+        sort: QuerySort<Member>,
+        members: List<Member>,
+    ): Result<List<Member>> {
+        val cidComponents = cid.split(':')
+
+        val result = domainImpl
+            .client
+            .queryMembers(cidComponents[0], cidComponents[1], offset, limit, filter, sort, members)
+            .execute()
+
+        if (result.isSuccess) {
+            domainImpl.repos.updateMembersForChannel(cid, result.data())
+        }
+
+        return result
+    }
+
+    private suspend fun queryMembersOffline(
+        cid: String,
+        sort: QuerySort<Member>,
+        offset: Int,
+        limit: Int,
+    ): Result<List<Member>> {
+        // retrieve from database
+        val membersFromDatabase = domainImpl
+            .repos
+            .selectMembersForChannel(cid)
+            .takeIf { it.isNotEmpty() }
+            ?.let { members ->
+                members
+                    .sortedWith(sort.comparator)
+                    .drop(offset)
+                    .let { afterDrop ->
+                        if (limit > 0) {
+                            afterDrop.take(limit)
+                        } else afterDrop
+                    }
+            } ?: emptyList()
+
+        return Result(membersFromDatabase)
```

Comment:
Why do you do all these checks? Kotlin built-in collection extensions do all this stuff already. You don't need to check if the collection is empty for applying drop operator. With a negative limit value, the take operator will throw an exception.

```suggestion
return domainImpl.repos
            .selectMembersForChannel(cid)
            .sortedWith(sort.comparator)
            .drop(offset)
            .take(limit)
            .let(::Result)
```

Reply:
Default take value of 0 will return an empty list, so some checks are needed.